### ISSUE TITLE
TypeError (actually more like a NameError?) in inferences.py

### ISSUE THREAD

**Issue by HenryDayHall** [eth-cscs/abcpy#83]
I am trying to start the process from a journal file read in from disk, and getting a TypeError from inferences.py.

So earlier in the code `accepted_cov_mats=None`, then in these lines (435-438);
```
                new_cov_mats = self.kernel.calculate_cov(self.accepted_parameters_manager)
                # Since each entry of new_cov_mats is a numpy array, we can multiply like this
                # accepted_cov_mats = [covFactor * new_cov_mat for new_cov_mat in new_cov_mats]
                accepted_cov_mats = self._compute_accepted_cov_mats(covFactor, accepted_cov_mats)
```
This throws a TypeError, as `None` is not valid for the second parameter of `self._compute_accepted_cov_mats`.
Once you go into `self._compute_accepted_cov_mats` the second parameter is named `new_cov_mats`.
Should the third line be uncommented, or the second parameter `accepted_cov_mats` actually be `new_cov_mats`? 

Best,
Henry

**Comment by LoryPack** [comment]
Hello, 

That is a bug of which we are already aware (see #78). The right way to fix it is to change `accepted_cov_mats` in line 438 to `new_cov_mats`. We are planning to have a new release soon which addresses that issue (probably some time next week); otherwise, you can install from source the develop branch, where that has already been fixed.

**Comment by HenryDayHall** [comment]
So I made the change, but the parameters from the journal file don't seem to be used.
Here is a graph of training progress over two sessions;
![training_issues](https://user-images.githubusercontent.com/12996763/104731748-1197a980-5734-11eb-81b7-7e6dab085348.png)

I appreciate that this would be easier to work on if I gave a minimal working (broken) example or pointed to a change that fixed it, I will try to update with one of those two soon.

**Comment by LoryPack** [comment]
Yes I know that is an issue with the code; that version did not use correctly the final threshold value which is stored in the journal file. We however fixed that on the develop branch, and planning to include that in the next release I mentioned. You can fix that by either installing the develop branch of abcpy from source as I suggested, or if you are working with the source code itself it should suffice to take the `inferences.py` file [there](https://github.com/eth-cscs/abcpy/blob/af7361f43a3570ded6406174653aa2fc9c9097b7/abcpy/inferences.py) and replace the one you are currently using (or even replace only the PMCABC class).

Please let me know if you are able to make it work,  and sorry about that.

**Comment by HenryDayHall** [comment]
Ah ok, sorry I am just informing you of issues that you already have noted. I will switch to the development branch to put a stop to that.

**Comment by HenryDayHall** [comment]
It's still not working unfortunately. I download inferences.py and imported it with a relative file path, so it's certainly the version you linked.
![abcpyissue](https://user-images.githubusercontent.com/12996763/104900755-5ad63c00-5974-11eb-8008-641bde35c0a5.png)

So looking at the sample function, this is where it should be picking up from where it left off;
```
        for aStep in range(steps):
            self.logger.debug("iteration {} of PMC algorithm".format(aStep))
            if aStep == 0 and journal_file is not None:
                accepted_parameters = journal.get_accepted_parameters(-1)
                accepted_weights = journal.get_weights(-1)

                if  hasattr(journal, "distances"):
                    # if restarting from a journal, use the previous distances to check determine a new epsilon
                    # (it if is larger than the epsilon_arr[0] provided here)
                    epsilon_arr[0] = np.max([np.percentile(journal.distances[-1], epsilon_percentile), epsilon_arr[0]])

                self.accepted_parameters_manager.update_broadcast(self.backend, accepted_parameters=accepted_parameters,
                                                                  accepted_weights=accepted_weights)
```
_Question 1;_ I'm not seeing why this is this in the loop, and perhaps understanding that would help me understand what else is happening. It can only ever run on the first step (`aStep==0`) so surely it could happen before the loop with just `if journal_file is not None`? 

_Question 2;_ at the end of this snippet on line 428 of the full file there is an update broadcast. Then the same thing happens again on 452. Does it need to happen twice?

_Question 3;_ I'm using the dummy backend, is there an attribute that I should be monitoring to understand when it 'forgets'?

**Comment by LoryPack** [comment]
Hello, sorry for replying to this a bit late. 

**First about your questions:**

Q1: you are right, that could also go before the loop starts; it is just a programming choice, nothing would change - except needing to check once every step whether `aStep==0`, which has however negligible cost with respect to all the rest of the algorithm. 

Q2: I think the reason why the first broadcast is needed is for the calculation of the kernel parameters which come immediately below to work: 
https://github.com/eth-cscs/abcpy/blob/af7361f43a3570ded6406174653aa2fc9c9097b7/abcpy/inferences.py#L428-L439

Q3: the dummy backend should not change the behavior of your code. 

**Regarding the reasons why it does not work in your case:**
The part of code which is responsible for restarting from a journal file in the same way as if the algorithm was running continuously is:

https://github.com/eth-cscs/abcpy/blob/af7361f43a3570ded6406174653aa2fc9c9097b7/abcpy/inferences.py#L423-L426

and the idea there is that the value of the threshold `epsilon` is updated as in the rest of the algorithm:

https://github.com/eth-cscs/abcpy/blob/af7361f43a3570ded6406174653aa2fc9c9097b7/abcpy/inferences.py#L475-L476

but when you call the `sample` method from scratch providing a journal file, it uses the first value passed in the `epsilon_init` attribute (which is in `epsilon_arr[0]'). 

Then one reason why your simulation may not be working as you expect (meaning that when restarting from the Journal, the distances should still decrease) is that when you call `sample` for the second time (starting from the saved Journal) you are using a too large value of `epsilon_init`, so that line 426 selects that as a threshold for the current step, rather than the one which would be determined as the `epsilon_percentile` of the distances of the previous iteration (the one stored in the Journal file). 

To understand better, please have a look at how we test that restarting gives the same result as running all steps at once: 

https://github.com/eth-cscs/abcpy/blob/af7361f43a3570ded6406174653aa2fc9c9097b7/tests/inferences_tests.py#L214-L231

You can see above that in line 222 we use a different value for epsilon_init with respect to the one that was provided for the first step.

**Comment by statrita2004** [comment]
This should be fixed in our new release 0.6.1. Thanks a lot for looking into it in details. I am closing this issue now.
